### PR TITLE
Adds decorated `@property` support, refs #1362

### LINKED ISSUE #1362: Decorated property not supported
``` python
class Foo(object):
    @property
    @some_decorator
    def foo(self):
        ...
```

is valid in Python 2, but MyPy complains `Decorated property not supported`.
Adding `type: ignore` does not ignore the error.

### PR DESCRIPTION
**Pull request by sobolevn**
This is not fully ready right now, but looks like the first test passes 🎉 

Closes #1362

### CODE REVIEW

**Comment by sobolevn** [comment]
Things TODO:
- [ ] Add support for decorated `@property.setter`
- [ ] Add better error message for `@overload` and `@property` used together. I don't plan to support this feature in this PR, but I guess overloded properties can be a thing later, since we can have different `self` types

**Comment by sobolevn** [comment]
Friendly ping to @unparalleled-js. You offered your help, so may you would have some time to review my code 🙂

File: mypy/checker.py [hunk 1]
```diff
@@ -3851,6 +3868,20 @@ def visit_decorator(self, e: Decorator) -> None:
             if e.type and not isinstance(get_proper_type(e.type), (FunctionLike, AnyType)):
                 self.fail(message_registry.BAD_CONSTRUCTOR_TYPE, e)
 
+    def get_decorator_fullname_and_object_type(
```
Comment (unparalleled-js) [inline mypy/checker.py:3871]: Is there a reason to not keep this method protected?

```suggestion
    def _get_decorator_fullname_and_object_type(
```

Comment (sobolevn) [reply mypy/checker.py:3871]: I 100% agree, but mypy does not use protected methods that much. So, almost everything is "public". I prefer to follow the style here 🙂

**Comment by unparalleled-js** [comment]
I tested it out for both `@property` and `@cached_property` and ran `mypy` and some other projects, as a manual regression.

I am not sure about the other test failures yet but I can keep diving in, this is my first time peaking around `mypy` but I am invested in this issue.

**Comment by sobolevn** [comment]
> I tested it out for both @property and @cached_property and ran mypy and some other projects, as a manual regression.

Thank you! Any interesting finds? 🙂 

> I am not sure about the other test failures

Test failures are about incorrect test fixture and explicit `Any` in it. I will fix it later today. It is not related to the feature itself (I hope).

> this is my first time peaking around mypy but I am invested.

🚀 👍

**Comment by sobolevn** [comment]
Ok, I have to admit that my `.property_decorator` hack didn't work. We have too many possible decorators that can go on top: `@final`, `@no_type_check`, `@custom_cache`, `@abtract`, etc.

Fixing one problem and introducing another one is not ideal.

---

New plan: `@property` is just a regular descriptor with `__get__` and `__set__` methods. I guess I should concentrate on making it work as it should.

**Pull request closed without merging**